### PR TITLE
[Semi-Modular] Adds TEG back, nerfs its power generation and resprites it 

### PR DESCRIPTION
**Pull request by Erol509**
## About The Pull Request

https://github.com/shiptest-ss13/Shiptest/pull/817

Visuals: https://github.com/yogstation13/Yogstation/pull/9195

TEG in this file uses different power generation, which requires some different setups for it to generate power. You cant just plug couple of thermomachines, set them to max with easy to make hypernoblium to make 200 MW of power, you can stack TEGS so it would even more power

Im using logistic function from Shiptest which will only start making power when temperatures are going to reach quite high, so you will need proper burn chamber or at least burn can. Its something different than pluging teg to few thermomachines and making more than SM multiplied by 100. 

/// A function that exponentially approaches a maximum value of L
/// k is the rate at which it approaches L, x_0 is the point where the function = 0
#define LOGISTIC_FUNCTION(L,k,x,x_0) (L/(1+(NUM_E**(-k*(x-x_0)))))

Also Im using different sprites for the TEG, I took them Yogstation, they have their own ambient light and pulse animations. I think its a really nice sprite. 

Teg also was added to techweb, and can now be made in the engineering protolathe. 

## How This Contributes To The Skyrat Roleplay Experience

Perhaps another engine for our modular engine PR, or just another project for our nice atmos engineers. TEG is tricky to set up, due to new pump limits when working with hight temperatures, but it can be worked with with special setups. 

## Proof of Testing

image
image

## Changelog

:cl:
add: Added back TEG to techwebs and engineering protolathe
image: Changed TEG sprites from YogStation. Credits to original author.
/:cl:

### CODE REVIEW

**Comment by Erol509** [comment]
@Useroth @Gandalf2k15

**Comment by Useroth** [comment]
i like it, may need a sync with master branch cause we had some teg related prs lately

**Comment by Erol509** [comment]
> i like it, may need a sync with master branch cause we had some teg related prs lately

Fixed CI, I might need to actually change ssvis overlay to mutable appearance since as they said on tgstation, they lower the server workload

**Comment by Erol509** [comment]
@Useroth Ugh...Its ready to test or merge.

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+					set_light(2,1,"#4F82FF")
+				else
+					set_light(2,1,"#FF3232")
+				SSvis_overlays.add_vis_overlay(src, icon, "circ-[mode?"cold":"hot"]", LIGHTING_ABOVE_ALL, ABOVE_LIGHTING_PLANE, dir)
+				SSvis_overlays.add_vis_overlay(src, icon, "circ-slow", LIGHTING_ABOVE_ALL, ABOVE_LIGHTING_PLANE, dir)
+
+/obj/machinery/atmospherics/components/binary/circulator/wrench_act(mob/living/user, obj/item/I)
+
+	if(!panel_open)
+		balloon_alert(user, "<span class='warning'>Open the panel first!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
		balloon_alert(user, span_warning("Open the panel first!"))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		return TRUE
+
+	if(generator)
+		balloon_alert(user, "<span class='warning'>Disconnect [generator] first!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
		balloon_alert(user, span_warning("Disconnect [generator] first!"))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		return TRUE
+
+	set_anchored(!anchored)
+	I.play_tool_sound(src)
+	balloon_alert(user, "<span class='notice'>You [anchored?"secure":"unsecure"] [src].</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
	balloon_alert(user, span_notice("You [anchored?"secure":"unsecure"] [src]."))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+	return list(turn(dir, 90), turn(dir, 270))
+
+/obj/machinery/atmospherics/components/binary/circulator/can_be_node(obj/machinery/atmospherics/target)
+	if(anchored)
+		return ..(target)
+	return FALSE
+
+/obj/machinery/atmospherics/components/binary/circulator/multitool_act(mob/living/user, obj/item/I)
+	if(generator)
+		balloon_alert(user, "<span class='warning'>Disconnect [generator] first!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
		balloon_alert(user, span_warning("Disconnect [generator] first!"))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+		return TRUE
+
+	mode = !mode
+	balloon_alert(user, "<span class='notice'>You set [src] to [mode?"cold":"hot"] mode.</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
	balloon_alert(user, span_notice("You set [src] to [mode?"cold":"hot"] mode."))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+/obj/machinery/atmospherics/components/binary/circulator/screwdriver_act(mob/user, obj/item/I)
+	if(..())
+		return TRUE
+	if(generator)
+		balloon_alert(user, "<span class='warning'>Disconnect the generator first!</span>")
+		return TRUE
+	panel_open = !panel_open
+	I.play_tool_sound(src)
+	balloon_alert(user, "<span class='notice'>You [panel_open?"open":"close"] the panel on [src].</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
	balloon_alert(user, span_notice("You [panel_open?"open":"close"] the panel on [src]."))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+	mode = !mode
+	balloon_alert(user, "<span class='notice'>You set [src] to [mode?"cold":"hot"] mode.</span>")
+	return TRUE
+
+/obj/machinery/atmospherics/components/binary/circulator/screwdriver_act(mob/user, obj/item/I)
+	if(..())
+		return TRUE
+	if(generator)
+		balloon_alert(user, "<span class='warning'>Disconnect the generator first!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
		balloon_alert(user, span_warning("Disconnect the generator first!"))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+/obj/machinery/atmospherics/components/binary/circulator/default_change_direction_wrench(mob/user, obj/item/I)
+	if(!..())
+		return FALSE
+	set_init_directions()
+	update_appearance()
+	return TRUE
+
+/obj/machinery/atmospherics/components/binary/circulator/crowbar_act(mob/user, obj/item/I)
+	if(anchored)
+		balloon_alert(user, "<span class='warning'>[src] is anchored!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
		balloon_alert(user, span_warning("[src] is anchored!"))
```

File: modular_skyrat/master_files/code/modules/power/circulator.dm [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+/obj/machinery/atmospherics/components/binary/circulator/verb/circulator_flip()
+	set name = "Flip"
+	set category = "Object"
+	set src in oview(1)
+
+	if(!ishuman(usr))
+		return
+	flipped = !flipped
+	balloon_alert(usr, "<span class='notice'>You flip [src].</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/circulator.dm]:
```suggestion
	balloon_alert(usr, span_notice("You flip [src]."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+/obj/machinery/power/thermoelectric_generator/wrench_act(mob/living/user, obj/item/I)
+	. = ..()
+
+	if(!panel_open) //connect/disconnect circulators
+		if(!anchored)
+			balloon_alert(user, "<span class='warning'>Anchor [src] before trying to connect the circulators!</span>")
+			return TRUE
+		else
+			if(hot_circ && cold_circ)
+				balloon_alert(user, "<span class='notice'>You start removing the circulators...</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
				balloon_alert(user, span_notice("You start removing the circulators..."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+				if(I.use_tool(src, user, 30, volume=50))
+					null_circulators()
+					update_appearance()
+					balloon_alert(user, "<span class='notice'>You disconnect [src]'s circulator links.</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
					balloon_alert(user, span_notice("You disconnect [src]'s circulator links."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+					playsound(src, 'sound/misc/box_deploy.ogg', 50)
+				return TRUE
+
+			balloon_alert(user, "<span class='notice'>You attempt to attach the circulators...</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
			balloon_alert(user, span_notice("You attempt to attach the circulators..."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+			if(I.use_tool(src, user, 30, volume=50))
+				switch(find_circulators())
+					if(0)
+						balloon_alert(user, "<span class='warning'>No circulators found!</span>")
+					if(1)
+						balloon_alert(user, "<span class='warning'>Only one circulator found!</span>")
+					if(2)
+						balloon_alert(user, "<span class='notice'>You connect [src]'s circulator links.</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
						balloon_alert(user, span_notice("You connect [src]'s circulator links."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+					if(3)
+						balloon_alert(user, "<span class='warning'>Both circulators are the same mode!</span>")
+				return TRUE
+
+	set_anchored(!anchored)
+	I.play_tool_sound(src)
+	if(!anchored)
+		null_circulators()
+	connect_to_network()
+	balloon_alert(user, "<span class='notice'>You [anchored?"secure":"unsecure"] [src].</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
	balloon_alert(user, span_notice("You [anchored?"secure":"unsecure"] [src]."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+/obj/machinery/power/thermoelectric_generator/screwdriver_act(mob/user, obj/item/I)
+	if(..())
+		return TRUE
+
+	if(hot_circ && cold_circ)
+		balloon_alert(user, "<span class='warning'>Disconnect the circulators first!</span>")
+		return TRUE
+	panel_open = !panel_open
+	I.play_tool_sound(src)
+	balloon_alert(user, "<span class='notice'>You [panel_open?"open":"close"] the panel on [src].</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
	balloon_alert(user, span_notice("You [panel_open?"open":"close"] the panel on [src]."))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+		if(circulators.mode == CIRCULATOR_HOT && !hot_circ)
+			hot_circ = circulators
+			circulators.generator = src
+
+/obj/machinery/power/thermoelectric_generator/wrench_act(mob/living/user, obj/item/I)
+	. = ..()
+
+	if(!panel_open) //connect/disconnect circulators
+		if(!anchored)
+			balloon_alert(user, "<span class='warning'>Anchor [src] before trying to connect the circulators!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
			balloon_alert(user, span_warning("Anchor [src] before trying to connect the circulators!"))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+					update_appearance()
+					balloon_alert(user, "<span class='notice'>You disconnect [src]'s circulator links.</span>")
+					playsound(src, 'sound/misc/box_deploy.ogg', 50)
+				return TRUE
+
+			balloon_alert(user, "<span class='notice'>You attempt to attach the circulators...</span>")
+			if(I.use_tool(src, user, 30, volume=50))
+				switch(find_circulators())
+					if(0)
+						balloon_alert(user, "<span class='warning'>No circulators found!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
						balloon_alert(user, span_warning("No circulators found!"))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+					if(1)
+						balloon_alert(user, "<span class='warning'>Only one circulator found!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
						balloon_alert(user, span_warning("Only one circulator found!"))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+					if(2)
+						balloon_alert(user, "<span class='notice'>You connect [src]'s circulator links.</span>")
+						playsound(src, 'sound/misc/box_deploy.ogg', 50)
+						return TRUE
+					if(3)
+						balloon_alert(user, "<span class='warning'>Both circulators are the same mode!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
						balloon_alert(user, span_warning("Both circulators are the same mode!"))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+	balloon_alert(user, "<span class='notice'>You [anchored?"secure":"unsecure"] [src].</span>")
+	update_appearance()
+	return TRUE
+
+/obj/machinery/power/thermoelectric_generator/screwdriver_act(mob/user, obj/item/I)
+	if(..())
+		return TRUE
+
+	if(hot_circ && cold_circ)
+		balloon_alert(user, "<span class='warning'>Disconnect the circulators first!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
		balloon_alert(user, span_warning("Disconnect the circulators first!"))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+	panel_open = !panel_open
+	I.play_tool_sound(src)
+	balloon_alert(user, "<span class='notice'>You [panel_open?"open":"close"] the panel on [src].</span>")
+	update_appearance()
+	return TRUE
+
+/obj/machinery/power/thermoelectric_generator/crowbar_act(mob/user, obj/item/I)
+
+	if(anchored)
+		balloon_alert(user, "<span class='warning'>[src] is anchored!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
		balloon_alert(user, span_warning("[src] is anchored!"))
```

File: modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+		return TRUE
+	else if(!panel_open)
+		balloon_alert(user, "<span class='warning'>Open the panel first!</span>")
```
Comment (Gboster-0) [inline modular_skyrat/master_files/code/modules/power/thermoelectric_generator.dm]:
```suggestion
		balloon_alert(user, span_warning("Open the panel first!"))
```

**Review by Gboster-0 — COMMENTED** [review]
span :3

**Review by Gandalf2k15 — APPROVED** [review]

**Pull request merged**